test/e2e/framework/util.go

```diff
@@ -143,6 +147,28 @@ func waitTimeoutForPodRunningInNamespace(c kubernetes.Interface, podName, namesp
 	return wait.PollImmediate(Poll, defaultTimeout, podRunning(c, podName, namespace))
 }
 
+func WaitForEndpoint(c kubernetes.Interface, namespace, name string) error {
+	return wait.PollImmediate(Poll, EndpointRegisterTimeout, endpointAvailable(c, namespace, name))
+}
+
+func endpointAvailable(c kubernetes.Interface, namespace, name string) wait.ConditionFunc {
+	return func() (bool, error) {
+		endpoint, err := c.CoreV1().Endpoints(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		if len(endpoint.Subsets) == 0 || len(endpoint.Subsets[0].Addresses) == 0 {
```

Comment:
What is an endpoints and how is this working?

Reply:
Look up the docs on how Kube services work

Reply:
TLDR: endpoints are where the endpoint information for services lives.  The endpoints controller populates them and keeps the list up to date.  This check basically says "make sure at least one pod is backing this service"